sdk/eventhub/azure-eventhubs/azure/eventhub/aio/_producer_client_async.py

```diff
@@ -0,0 +1,179 @@
+            async with self._client_lock:
+                if not self._producers:
+                    num_of_producers = len(await self.get_partition_ids()) + 1
+                    self._producers = [None] * num_of_producers
+                    for i in range(num_of_producers):
+                        self._producers_locks.append(asyncio.Lock())
+                        # self._producers_locks = [asyncio.Lock()] * num_of_producers
+
+    async def send(self, event_data: Union[EventData, EventDataBatch, Iterable[EventData]],
+            *, partition_key: Union[str, bytes] = None, partition_id: str = None, timeout: float = None) -> None:
+        """Sends an event data and blocks until acknowledgement is received or operation times out.
+
+        :param event_data: The event(s) to be sent. It can be an EventData object, an iterable of EventData objects
+            or an EventDataBatch, which contains some EventData objects. EventDataBatch is highly recommended to group
+            as many EventData objects as possible in a single send call. This will maximize the send throughput.
+            A send operation will send all EventData objects to one partition.
+            Multiple send operations will round-robin send to partitions alternately.
+            Round-robin is the recommended and the default behavior of send
+            unless you set partition_id or partition_key.
+        :param partition_id: event_data will be sent to this partition. This is not recommended because it may
+         cause unbalanced number of event data across partitions.
+        :param partition_key: With the given partition_key, event_data will land to
+         a particular partition but unknown until the send completes. Two send operations with the same partition_key
+         will go to the same partition even though which the partition is unknown while before send.
+        :param timeout: The maximum wait time to send the event data.
+         Default value is 60 seconds. If set to 0, there will be no timeout.
+
+        :raises: ~azure.eventhub.AuthenticationError, ~azure.eventhub.ConnectError, ~azure.eventhub.ConnectionLostError,
+                ~azure.eventhub.EventDataError, ~azure.eventhub.EventDataSendError, ~azure.eventhub.EventHubError
+        :return: None
+        :rtype: None
+
+        Example:
+            .. literalinclude:: ../samples/test_examples_eventhub_async.py
+                :start-after: [START eventhub_producer_client_send_async]
+                :end-before: [END eventhub_producer_client_send_async]
+                :language: python
+                :dedent: 4
+                :caption: Asynchronously sends an event data
+
+        """
+
+        await self._init_locks_for_producers()
+
+        producer_index = int(partition_id) if partition_id is not None else -1
+        if self._producers[producer_index] is None or self._producers[producer_index]._closed:  # pylint:disable=protected-access
```

Comment:
If we are using the `_closed` attribute, we should make it public. We want to avoid protected access warnings since they become a maintenance burden (it is no longer safe to assume that things with internal names are indeed internal otherwise)

Reply:
not blocking. let's leave it later. This is small change but very easy to make bugs